### DISCUSSION TITLE
Production build will not connect to MongoDB

### DISCUSSION THREAD

**Discussion by omar-enrique** [strapi/strapi#5768] [Legacy Topics]
**Describe the bug**
This could easily be something I am doing wrong, but I have read past discussions on this topic, such as #4311 and #636, and I have still not found a solution to this problem.  When I run my Strapi CMS locally with "yarn run start," Strapi serves locally at port 1337 with no problems connecting to my MongoDB.  I am easily able to interface with my mongodb data, showing the connection string works.  When I deployed Strapi on Heroku however, using "git push heroku master," strapi will fail to start with the following error:

```
2020-04-10T16:49:04.971534+00:00 app[web.1]: > portfolio-mongo-cms@0.1.0 start /app
2020-04-10T16:49:04.971534+00:00 app[web.1]: > strapi start
2020-04-10T16:49:04.971535+00:00 app[web.1]:
2020-04-10T16:49:37.139728+00:00 app[web.1]: [2020-04-10T16:49:37.138Z] debug ⛔️ Server wasn't able to start properly.
2020-04-10T16:49:37.140759+00:00 app[web.1]: [2020-04-10T16:49:37.140Z] error Error connecting to the Mongo database. Server selection timed out after 30000 ms
2020-04-10T16:49:37.151045+00:00 app[web.1]: npm ERR! code ELIFECYCLE
2020-04-10T16:49:37.151355+00:00 app[web.1]: npm ERR! errno 1
2020-04-10T16:49:37.152214+00:00 app[web.1]: npm ERR! portfolio-mongo-cms@0.1.0 start: `strapi start`
2020-04-10T16:49:37.152310+00:00 app[web.1]: npm ERR! Exit status 1
2020-04-10T16:49:37.152428+00:00 app[web.1]: npm ERR!
2020-04-10T16:49:37.152513+00:00 app[web.1]: npm ERR! Failed at the portfolio-mongo-cms@0.1.0 start script.
2020-04-10T16:49:37.152581+00:00 app[web.1]: npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

**Steps to reproduce the behavior**
I followed the Strapi documentation for deploying to a Heroku app.  I ran `heroku login` in my root directory for my strapi app, and added the heroku remote to point to my heroku application.  I added everything I wanted to add, leaving out node_modules and including yarn.lock, and running `git push heroku master` to start the build process on the production side.  The build finishes fine without errors and the deployment succeeds, but when I navigate to the URL, Strapi does not show up.  The heroku logs show the error that I displayed above.

**Expected behavior**
My strapi app connects to my MongoDB cluster the way it does when I run strapi locally.

**Code snippets**
Here is my development database.json:
```
{
  "defaultConnection": "default",
  "connections": {
    "default": {
      "connector": "mongoose",
      "settings": {
        "client": "mongo",
        "database": "test",
        "uri": "mongodb+srv://username:<password>@omar-portfolio-yuqdd.mongodb.net/test", 
        "host": "omar-portfolio-yuqdd.mongodb.net/test?retryWrites=true&w=majority",
        "srv": true,
        "port": 27017,
        "username": "",
        "password": ""
      },
      "options": {
        "authenticationDatabase": "",
        "ssl": true
      }
    }
  }
}
```

And this is my production database.json:
```
{
  "defaultConnection": "default",
  "connections": {
    "default": {
      "connector": "mongoose",
      "settings": {
        "client": "mongo",
        "database": "test",
        "uri": "mongodb+srv://username:<password>@omar-portfolio-yuqdd.mongodb.net/test", 
        "host": "omar-portfolio-yuqdd.mongodb.net/test?retryWrites=true&w=majority",
        "srv": true,
        "port": 27017,
        "username": "",
        "password": ""
      },
      "options": {}
    }
  }
}
```

**System**
- Node.js version: v10.15.3
- yarn version:  1.22.0
- Strapi version: 3.0.0-beta.18.8
- Database: MongoDB
- Local Operating system is Windows 10, Production version is hosted on Heroku

**Additional context**

**Comment by ziamohseni** [comment]
I had the same problem, but at last I did it :)
Follow this article on medium when installing on your local
https://medium.com/@firstsquares/local-strapi-cloud-mongodb-atlas-cc65288f0dee

then when you want to deploy, copy the database from development env and paste it in both production and staging env. 

and then deploy to heroku.

**Comment by yapnel** [comment]
I am also facing the same issue. working locally but deploying to cloud foundry I get the same error as you.

**Comment by derrickmehaffy** [comment]
Your database configs are wrong, here is a sample using atlas:

https://gist.github.com/derrickmehaffy/aea3c635e0b9704b14fd1693fc56fea2

**Comment by LuizCristino** [comment]
Tried any combination of settings I could find on stackoverflow and github issues none of them worked with Atlas. 

Also checked my connection string using mongo cli both versions 2.2 and 3.0 connection strings worked

database.json
```json
{
  "defaultConnection": "default",
  "connections": {
    "default": {
      "connector": "mongoose",
      "settings": {
        "uri": "mongodb://user:password@cluster0-shard-00-00-4dhof.mongodb.net:27017,cluster0-shard-00-01-4dhof.mongodb.net:27017,cluster0-shard-00-02-4dhof.mongodb.net:27017/test?ssl=true&replicaSet=Cluster0-shard-0&authSource=admin&retryWrites=true&w=majority"
      },
      "options": {
        "authenticationDatabase": "admin",
        "ssl": true
      }
    }
  }
}

```

**Comment by omar-enrique** [comment]
**My Solution: The problem was fixed for me!**
Alright so I tried the solution from the medium article @moh3ni posted and it worked. I did not re-create my strapi project, I just changed the development, production, and staging database.json files to be the exact same as each other. This is what they look like now:
```
{
  "defaultConnection": "default",
  "connections": {
    "default": {
      "connector": "mongoose",
      "settings": {
        "client": "mongo",
        "database": "test",
        "host": "omar-portfolio-yuqdd.mongodb.net",
        "srv": true,
        "port": 27017,
        "username": username,
        "password": password
      },
      "options": {
        "authenticationDatabase": "",
        "ssl": true
      }
    }
  }
}
```
@LuizCristino, if your mongodb cluster is SRV then all you would have to do is put what I put in host, which would just be the .net URL after the @ symbol in your connection string.  This would umbrella all the shards in your MongoDB cluster. My username and password is the user I created in my Database Access settings.

I copy and pasted this to all three database.json files and ran "git push heroku master" and it worked.

**Reply by marshyon** [reply]
this works for me too, thanks for posting this !

so I think the docs feature postgres and for mongodb atlas the bits that may be missing are the 'options' section to set ssl to true, the 'srv' to true and 'client' to be mongo. I had already added 'connector' to be mongoose but was missing the others.

**Comment by LuizCristino** [comment]
So I tried a `database.json` with same structure. Now bootsrap function fails to execute. Something that doesn't happen when I use mongodb locally.

I cannot understand why it connect and works locally but cannot connect to atlas. 

I will mention here that I gave access to my ip address I can connect to database using mongo command line. 

```
$ strapi start
[2020-04-11T19:25:29.230Z] warn The bootstrap function is taking unusually long to execute (3500 miliseconds).
[2020-04-11T19:25:29.231Z] warn Make sure you call it?
[2020-04-11T19:25:29.248Z] warn The bootstrap function is taking unusually long to execute (3500 miliseconds).
[2020-04-11T19:25:29.249Z] warn Make sure you call it?
[2020-04-11T19:26:06.199Z] error Bootstrap function in plugin "content-manager" failed
[2020-04-11T19:26:06.200Z] error Error: read ECONNRESET
    at TLSWrap.onStreamRead (internal/stream_base_commons.js:200:27)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

**Comment by omar-enrique** [comment]
@LuizCristino did you make sure to white list all IP addresses? If you only allow your local IP address, your production server will not be able to connect to the MongoDB.

As for the bootstrap error, that looks like an unrelated issue. You most likely have the right Database.json now. Did you add new bootstrap components to your strapi project that were not already build in to the boilerplate?

**Comment by LuizCristino** [comment]
I did the following steps

1. Created an new strapi app with custom settings
2. Setted all the config to mongo database hosted on my localhost
3. Run yarn/npm develop to testing it would work
4. Created my project and cluster on atlas
5. Created user with read and write permission. After while I tried to change to admin to check it
![image](https://user-images.githubusercontent.com/33782549/79056235-a8c25480-7c2a-11ea-9082-f95d04de3e97.png)

6. Whitelisted all ips (just for sake of testing)
![image](https://user-images.githubusercontent.com/33782549/79056227-89c3c280-7c2a-11ea-9a1a-96c27e0166d5.png)

For short I'm trying to connect a blank app

**Comment by omar-enrique** [comment]
@LuizCristino Forgive me if I am misunderstanding, but if you are locally hosting your MongoDB, you are going to need to migrate that MongoDB to the Cloud for it to work with a production build.

**Comment by LuizCristino** [comment]
And that is exactly where the problem is. Strapi isn't working with Atlas. I was stuck two days trying to migrate my local to a remote service (Atlas).

Now I'm migrating my backend to Laravel, could not make Strapi connect to Atlas correctly

**Comment by yapnel** [comment]
Your order of steps are incorrect to use Atlas with a new strapi project. You should setup your Atlas first, then, a new strapi project with Atlas connection details. 

What you have done is created strapi metadata in your local mongo then you change the connection to Atlas that has no strapi metadata in it.

**Comment by LuizCristino** [comment]
@yapnel Atlas was configured prior strapi attempt to connection. I could connect to atlas via mongo shell and an empty nodejs project.

**Comment by lauriejim** [comment]
Hello there!
Here is a quick video about using Strapi and MongoDB Atlas - [video](https://www.youtube.com/watch?v=HHnRsZwFmR8&list=PL7Q0DQYATmvhlHxHqfKHsr-zFls2mIVTi&index=7&t=0s)

And here in the documentation, there is a guide of configuring and using MongoDB Atlas in your project - [documentation](https://strapi.io/documentation/3.0.0-beta.x/guides/databases.html#mongodb-installation)
